forc-util/src/tx_utils.rs

```diff
@@ -0,0 +1,315 @@
+    match program_abi {
+        ProgramABI::Fuel(program_abi) => decode_fuel_vm_log_data(log_id, log_data, program_abi),
+        _ => Err(anyhow::anyhow!(
+            "only Fuel VM is supported for log decoding"
+        )),
+    }
+}
+
+pub fn decode_fuel_vm_log_data(
+    log_id: &str,
+    log_data: &[u8],
+    program_abi: &fuel_abi_types::abi::program::ProgramABI,
+) -> anyhow::Result<DecodedLog> {
+    let program_abi =
+        fuel_abi_types::abi::unified_program::UnifiedProgramABI::from_counterpart(program_abi)?;
+
+    // Create type lookup (id, TypeDeclaration)
+    let type_lookup = program_abi
+        .types
+        .iter()
+        .map(|decl| (decl.type_id, decl.clone()))
+        .collect::<HashMap<_, _>>();
+
+    let logged_type_lookup: HashMap<_, _> = program_abi
+        .logged_types
+        .iter()
+        .flatten()
+        .map(|logged_type| (logged_type.log_id.as_str(), logged_type.application.clone()))
+        .collect();
+
+    let type_application = logged_type_lookup
+        .get(&log_id)
+        .ok_or_else(|| anyhow::anyhow!("log id is missing"))?;
+
+    let abi_decoder = ABIDecoder::default();
+    let param_type = ParamType::try_from_type_application(type_application, &type_lookup)?;
+    let decoded_str = abi_decoder.decode_as_debug_str(&param_type, log_data)?;
+    let decoded_log = DecodedLog { value: decoded_str };
+
+    Ok(decoded_log)
+}
+
+pub struct RevertPosition {
+    pub pkg: String,
+    pub file: String,
+    pub line: u64,
+    pub column: u64,
+}
+
+// TODO: Move `RevertInfo` and related types to `fuel-abi-types` crate.
+//       We temporarily keep it here to get the support for `panic` expression in `forc test` ASAP,
+//       without waiting for the next `fuel-abi-types` release.
+
+/// Information about a revert that occurred during a transaction execution.
+pub struct RevertInfo {
+    pub revert_code: u64,
+    pub kind: RevertKind,
+}
+
+pub enum RevertKind {
+    /// This is the most general kind of a revert, where we only know the revert code.
+    /// E.g., reverts caused by `__revert` calls.
+    RawRevert,
+    /// Reverts caused by known functions, like, e.g., `assert` or `require`, that provide known error signals.
+    /// For such reverts, we can provide the error message.
+    KnownErrorSignal { err_msg: String },
+    Panic {
+        err_msg: Option<String>,
+        err_val: Option<String>,
+        pos: RevertPosition,
+    },
+}
+
+impl RevertInfo {
+    pub fn raw_revert(revert_code: u64) -> Self {
+        Self {
+            revert_code,
+            kind: RevertKind::RawRevert,
+        }
+    }
+
+    pub fn new(
+        revert_code: u64,
+        program_abi: Option<&fuel_abi_types::abi::program::ProgramABI>,
+        logs: &[fuel_tx::Receipt],
+    ) -> Self {
+        /// Types that implement the `std::marker::Error` trait, and whose instances
+        /// can be used as arguments to the `panic` expression.
+        enum ErrorType {
```

Comment:
if possible would be great to have unit tess for each of these variants in this file

Reply:
The tests for all the supported variants of the `Error` trait, effectively testing this functionality, are done in the `should_pass/language/panic_expression/panic_handling_in_unit_tests`, added in this PR. By far most of the compiler tests are e2e tests, that's why I put them there, instead of having unit tests here. To test the same functionality in unit tests would require a convenient infrastructure for creating consistent `ProgramABI` instances in tests, which we don't have now.

While I am all for Rust unit tests, in this case they would test what's already tested, and require quite of an effort to manually create required `ProgramABI` so I would rather skip it.

Reply:
makes sense 👍